programs/cltester/dwarf.cpp

```diff
@@ -0,0 +1,1812 @@
+      });
+
+      scan_custom(stream, [&](auto& section, const auto& name) {
+         if (name == ".debug_info")
+            dwarf::parse_debug_info(result, section.data);
+      });
+
+      std::sort(result.locations.begin(), result.locations.end());
+      std::sort(result.abbrev_decls.begin(), result.abbrev_decls.end());
+      if (show_wasm_loc_summary)
+         for (auto& loc : result.locations)
+            fprintf(stderr, "loc  [%08x,%08x) %s:%d\n", loc.begin_address, loc.end_address,
+                    result.files[loc.file_index].c_str(), loc.line);
+      if (show_wasm_subp_summary)
+         for (auto& p : result.subprograms)
+            fprintf(stderr, "subp %d [%08x,%08x) size=%08x %6s %s\n",
+                    int(&p - &result.subprograms[0]), p.begin_address, p.end_address,
+                    p.end_address - p.begin_address, p.parent ? "inline" : "",
+                    p.demangled_name.c_str());
+      return result;
+   }  // get_info_from_wasm
+
+   const char* info::get_str(uint32_t offset) const
+   {
+      eosio::check(offset < strings.size(), "string out of range in .debug_str");
+      return &strings[offset];
+   }
+
+   const location* info::get_location(uint32_t address) const
+   {
+      auto it = std::upper_bound(locations.begin(), locations.end(), address,
+                                 [](auto a, const auto& b) { return a < b.begin_address; });
+      if (it != locations.begin() && address < (--it)->end_address)
+         return &*it;
+      return nullptr;
+   }
+
+   const abbrev_decl* info::get_abbrev_decl(uint32_t table_offset, uint32_t code) const
+   {
+      auto key = std::pair{table_offset, code};
+      auto it = std::lower_bound(abbrev_decls.begin(), abbrev_decls.end(), key,
+                                 [](const auto& a, const auto& b) { return a.key() < b; });
+      if (it != abbrev_decls.end() && it->key() == key)
+         return &*it;
+      return nullptr;
+   }
+
+   const subprogram* info::get_subprogram(uint32_t address) const
+   {
+      auto it = std::upper_bound(subprograms.begin(), subprograms.end(), address,
+                                 [](auto a, const auto& b) { return a < b.begin_address; });
+      if (it != subprograms.begin() && address < (--it)->end_address)
+         return &*it;
+      return nullptr;
+   }
+
+   enum jit_actions : uint32_t
+   {
+      jit_noaction = 0,
+      jit_register_fn,
+      jit_unregister_fn
+   };
+
+   struct jit_code_entry
+   {
+      jit_code_entry* next_entry = nullptr;
+      jit_code_entry* prev_entry = nullptr;
+      const char* symfile_addr = nullptr;
+      uint64_t symfile_size = 0;
+   };
+
+   struct jit_descriptor
+   {
+      uint32_t version = 1;
+      jit_actions action_flag = jit_noaction;
+      jit_code_entry* relevant_entry = nullptr;
+      jit_code_entry* first_entry = nullptr;
+   };
+}  // namespace dwarf
+
+extern "C"
+{
+   void __attribute__((noinline, optnone)) __jit_debug_register_code(){};
```

Comment:
gcc 10 is optimizing out the call.

Reply:
dang it

Reply:
The gcc documentation of noinline suggests `asm("")`

Reply:
There were 2 issues with gcc:
* 10 optimized out the call
* 9 and 10 produced `*** longjmp causes uninitialized stack frame ***: terminated`

I added workarounds for both.